src/daisyworld/daisyworld/World.java

```diff
@@ -183,26 +197,59 @@ private void updateByDirection(Patch[][] newPatches, int i, int j, Directions di
                 newPatches[i][target].setTemp(newPatches[i][target].getTemp() + diff);
+                // set new patch soil quality
+                if (newPatches[i][target].getSoilQuality() == 0.0) {
+                    newPatches[i][target].setSoilQuality(0.0);
+                } else {
+                    newPatches[i][target].setSoilQuality(newPatches[i][target].getSoilQuality() + soilQualityDiff);
+                }
                 break;
             case DOWN:
                 target = j == size - 1 ? 0 : j + 1;
-                diff = (source.getTemp() - patches[i][target].getTemp()) / 2;
+                diff = (source.getTemp() - patches[i][target].getTemp()) * Params.DIFFUSION_RATE;
+                // calculation of soil quality difference
+                soilQualityDiff = (source.getSoilQuality() - patches[i][target].getTemp()) * Params.DIFFUSION_RATE;
                 newPatches[i][target].setTemp(newPatches[i][target].getTemp() + diff);
+                // set new patch soil quality
+                if (newPatches[i][target].getSoilQuality() == 0.0) {
+                    newPatches[i][target].setSoilQuality(0.0);
+                } else {
+                    newPatches[i][target].setSoilQuality(newPatches[i][target].getSoilQuality() + soilQualityDiff);
+                }
                 break;
             case LEFT:
                 target = i == 0 ? size - 1 : i - 1;
-                diff = (source.getTemp() - patches[target][j].getTemp()) / 2;
+                diff = (source.getTemp() - patches[target][j].getTemp()) * Params.DIFFUSION_RATE;
+                // calculation of soil quality difference
+                soilQualityDiff = (source.getSoilQuality() - patches[target][i].getTemp()) * Params.DIFFUSION_RATE;
                 newPatches[target][j].setTemp(newPatches[target][j].getTemp() + diff);
+                // set new patch soil quality
+                if (newPatches[target][i].getSoilQuality() == 0.0) {
+                    newPatches[target][i].setSoilQuality(0.0);
+                } else {
+                    newPatches[target][i].setSoilQuality(newPatches[target][i].getSoilQuality() + soilQualityDiff);
+                }
                 break;
             case RIGHT:
                 target = i == size - 1 ? 0 : i + 1;
-                diff = (source.getTemp() - patches[target][j].getTemp()) / 2;
+                diff = (source.getTemp() - patches[target][j].getTemp()) * Params.DIFFUSION_RATE;
+                // calculation of soil quality difference
+                soilQualityDiff = (source.getSoilQuality() - patches[target][i].getTemp()) * Params.DIFFUSION_RATE;
                 newPatches[target][j].setTemp(newPatches[target][j].getTemp() + diff);
+                // set new patch soil quality
+                if (newPatches[target][i].getSoilQuality() == 0.0) {
+                    newPatches[target][i].setSoilQuality(0.0);
```

Comment:
Seems to me the soil quality will never be 0 unless params.diffusion rate exceeds 1.0, which should not happen. Even if it happens, the soilQuality has no checks for being non-negative.

Reply:
I gonna correct it in the next version, using an extra parameter.